_posts/2022-05-11-backpressure.md

```diff
@@ -0,0 +1,203 @@
+---
+title: "How to process your data the resilient way with back pressure"
+description: "Alex explains to use back pressure with your OpenFaaS functions to make sure all your data gets processed."
+date: 2022-05-12
+image: /images/2022-react-app/pen-design.jpg
+categories:
+- architecture
+- casestudy
+- data
+- dataprocessing
+- bigdata
+author_staff_member: alex
+dark_background: true
+
+---
+
+Alex explains to use back pressure with your OpenFaaS functions to make sure all your data gets processed.
+
+As I've spoken to customers and community users of OpenFaaS, I've noticed a common theme. Users want to process large amounts of data through a number of functions and get results out the other end.
+
+The goal is to process a dataset as quickly as possible, without losing any of the records.
+
+I've seen friction in these areas:
+
+* timeouts being met on the gateway, queue-worker, or function
+* errors manifesting in functions that were not present under light load testing
+* dropped messages due to over saturating a function vs. how many requests it can process concurrently
+* being unsure of where to look first when problems arise
+
+I'll set out to explain why setting limits is important in a distributed system, along with a pattern called "back pressure" which can be used to ensure all of your requests get processed without losing data.
+
+## Why do we want limits?
+
+I started creating OpenFaaS in 2016 for three primary reasons: I wanted to run code in any language or runtime, be able to deploy it on any server or cloud, and to be able to set my own limits for timeouts, function size, and event sources for integrations.
+
+OpenFaaS gives you the ability to pick and choose a timeout that makes sense for your workload, instead of being fixed at 60 seconds. For asynchronous requests, there's a default limit for the payload size, however you can also change that.
+
+So if we can pick our limit, why even have one?
+
+Limits are important for understanding how a system will behave and how it can break, and even when we don't set limit explicitly, there may be some implicit ones which we are not aware of.
+
+Let's say that you configured a function to a limit of 1GB of RAM available for processing requests. You then load it with 1,000,000 requests. It's likely that it will break in correlation to the [ulimit](https://ss64.com/bash/ulimit.html) which is set up on the host. ulimits control how many processes can be started and how many files can be opened.
```

Comment:
`ulimits control`
to
`The ulimit values control`

to shift the ulimit lowercase into the sentence.